Lua implementation for l3intarray

zauguin opened the pull request:
An attempt to fix #788 and also an experiment about switching in Lua implementations for existing modules.



## TODO:
  - [x] The current system for defining the commands works with `luatexbase.new_luafunction`, which isn't available in ConTeXt. Add an abstraction for this allocator. (Related to #582)
  - [x] Fix the test. (The tests currently pass because the LaTeX2e version does not load ltluatex early enough, so the code is disabled. (See https://github.com/latex3/latex2e/pull/368) When testing with adapted 2e version, the test mostly work except for differences in behavior because of different limits.)
  - [x] Add a system for preserving intarrays created during format building for use in regular runs. An initial draft is part of this PR, but it's not used yet.

zauguin commented:
The older version was about three times slower (!) than the fontdimen version, the new version is much faster: For setting entries, we are about 30% faster than the fontdimen variant, about 10% faster for getting values and quite a bit slower (80%) for the kernel internal versions with less checking.

The non fontdimen verison has the additional advantage that it does not limit how big th eintegers are allowed to become.
(I guess that there is no big chance to get competitive performance for the kernel version because the fontdimen version does basically the same thing, except that it uses optimized C code instead of Lua...)

josephwright commented on l3kernel/l3luatex.dtx:
```diff
@@ -677,6 +677,84 @@ end
+%
+% We provide some kernel-internal helpers for this. They will only be available if
+% \texttt{luatexbase} is available. This is not a big restriction though, because
+% ConTeXt (which does not use \texttt{luatexbase}) does not load expl3 in the format.
```
'Con\TeX{}t', `\pkg{expl3}`

blefloch replied:
Also `forat` should be `format` a few lines above.

josephwright commented on l3kernel/l3luatex.dtx:
```diff
@@ -677,6 +677,84 @@ end
+
+    function register_luadata(name, func)
+      if luadata[name] then
+        error(string.format("LaTeX3 error: data name %q already in use", name))
```
I think we are just using 'LaTeX error' nowadays

josephwright commented on l3kernel/l3luatex.dtx:
```diff
@@ -677,6 +677,84 @@ end
+        local str = "return {"
+        for i=1, #luadata_order do
+          local name = luadata_order[i]
+          str = string.format('%s[%q]=%s,', str, name, luadata[name]())
```
Local copy of `string.format`?

josephwright commented on l3kernel/l3luatex.dtx:
```diff
@@ -677,6 +677,84 @@ end
+        end
+        lua.bytecode[register] = assert(load(str .. "}"))
+      end
+    end, "latex3.luadata")
```
Are we using `latex3` for this? I thought just `ltx`?

josephwright reviewed (commented):
Some queries - nothing major so merging then looking at them is fine.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+%   This has been chosen over a more conventional order to provide a delimiter for the numbers.
+%    \begin{macrocode}
+%<*lua>
+luacmd('@@_gset:wNwF', function()
```
I'm with @zauguin on this one, but my opinion was a minority when we decided a few years back, so let's do what Joseph suggests.

zauguin replied:
I think it would make sense to keep the `F` though to indicate that it behaves like a conditional, but I can drop that too if that's preferred.

blefloch replied:
Yes, `wF` should be fine and much clearer than just `w`.

zauguin commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+%    \begin{macrocode}
+%<*tex>
+\cs_if_exist:NTF \@@_gset_count:Nw
+{
```
I'll change this last since indenting the whole file makes working with diffs a pain.

zauguin replied:
Should be fixed.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+% \end{macro}
+%
+% \begin{macro}{\intarray_gzero:N, \intarray_gzero:c}
+%   Set the appropriate \tn{fontdimen} to zero.  No bound checking
+%   needed.
```
Stale comment.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+%   suitable for \cs{int_value:w}.
+%    \begin{macrocode}
+%<*lua>
+luacmd('@@_item:wNF', function()
+  local i = scan_int()
+  local t = @@_table()
+  local item = t[i]
+  if item then
+    put_next(use_none)
+  else
+    tex.l_@@_bad_index_int = i
+    put_next(use_i)
+  end
+  sprint(-2, item or 0)
+end, 'global')
```
This seems to leave something like `1234\use_none:n{false code}` in the input stream, instead of actually expanding the `\use_none:n` and getting just `1234`.  This matters in things like `\exp_after:Nf \iow_term:n { \intarray_item:Nn \g_tmpa_intarray { 12 } }`.

zauguin commented:
@josephwright Thanks. I addressed most of the remarks. There is one question left about `\@@_gset:wF` vs `\@@_gset_aux:w` and then I'll indent everything and merge :)

blefloch commented on l3kernel/testfiles/m3intarray001.luatex.tlg:
```diff
@@ -22,26 +22,19 @@ This is a coding error.
  ...                                              
 l. ...  }
-An attempt was made to access or store data at position 13 of the array
+An attempt was made to access or store data at position 0 of the array
```
This change is surprising and suggests that the LuaTeX version is losing track of the index number.

blefloch commented on l3kernel/testfiles/m3intarray001.luatex.tlg:
```diff
@@ -114,11 +75,11 @@ TEST 5: Zeroing
 0,0,0,0,0,0,0,20,0,0
 The integer array \g_testd_intarray contains 10 items:
->  0, 0, 0, 0, 0, 0, 0, 20, 0, 0.
+> 0, 0, 0, 0, 0, 0, 0, 20, 0, 0.
```
Ideally the extra space would be avoided in the non-Lua case?

zauguin replied:
I agree and have a patch for that, but it's a bit hacky and I wanted to keep it separate from this PR since it's rather unrelated.

blefloch replied:
Sounds good.

blefloch reviewed (commented):
I'm not really done but I have to stop.

Skillmon commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+  {
+    \__kernel_chk_if_free_cs:N #1
+    \int_gincr:N \g_@@_table_int
+    \cs_gset:Npx #1 { \s_@@ \int_use:N \g_@@_table_int \c_space_tl }
```
For kernel internal code there is the (marginally) faster `\__kernel_tl_set:Nx`.

zauguin replied:
Could you help me with an issue here? I tried changing this to `\__kernel_tl_gset:Nx` which complained about the variable not being declared. Is there some way to indicate to the debugging system to not check this instance?

Skillmon replied:
Haven't thought of the debug-redefinition... In that case maybe use `\cs_gset_nopar:Npx` instead of `\cs_gset:Npx`. Not sure whether this would be the correct style.

blefloch replied:
Ah, I forgot about the debug code, sorry.  Using `\cs_gset_nopar:Npx` seems the best then, that's what's used in `\tl_const:Nx`.  We cannot use `\tl_const:Nx` directly because the variable name starts with `g_` not `c_`.

zauguin replied:
Thanks, with `_nopar` it works fine.

zauguin commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+end
+
+local scan_argument = token.scan_argument
+local @@_table do
```
Lua doesn't care and I like doing this as a convention to indicate that the block is intended to initialize the local variable.
But we can split them, it's not really important.

blefloch replied:
I knew Lua doesn't care; I didn't know about this convention, thanks.

zauguin commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+  use_i = nil
+else
+  use_none = token.create'use_none:n'
+end
```
That happens by default. Since the Lua modules get combined into one `expl3.lua` and then get loaded as one of the first steps, `\use:n` will almost certainly be undefined at this point. By itself that wouldn't be an issue since the token references the command sequence ad not it's definition, but if the document never contained a `\use:n` token before then LuaTeX will not yet have added it to it's hash table and `token.create` returns a useless marker.

We could avoid the test here by  always assuming that they are undefined, but trying to assign values to them makes it IMO more obvious that these will end up being their values.

blefloch replied:
Looking at the order, it seems indeed that `ltluatex` and all the lua code is loaded before `ltexpl` and all the `expl3` code.  This means that the lines we're talking about never do anything.  I suggest removing these lines and simply adding a comment saying that `use_i` and `use_none` will eventually mean `\use:n` and `\use_none:n`.  Alternatively, to find a way to have `\use:n` and `\use_none:n` in the token stream before that point, possibly with `token.create'csname'` and so on.

zauguin replied:
> This means that the lines we're talking about never do anything

That's not quite true either because the Lua code is reloaded at `\everyjob` and at that point they are defined. But I'll find some other approach.

zauguin replied:
The new version introduces the tokens explicitly.

zauguin commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+%    \begin{macrocode}
+\msg_new:nnn { kernel } { negative-array-size }
+  { Size~of~array~may~not~be~negative:~#1 }
+%    \end{macrocode}
```
Thanks, I moved it into the common part.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+    tables[i] = table
+    return table
+  end
+  if register_luadata then
+    register_luadata('@@', function()
+      local t = "{[0]={},"
+      for i=1, #tables do
+        t = string.format("%s{%s},", t, table.concat(tables[i], ','))
+      end
+      return t .. "}"
+    end)
+  end
```
Ok, sorry, I was misled by the fact that here `table` refers to the built-in, not to your local variable.

zauguin replied:
I renamed the variable to hopefully make this less confusing and added some remarks about register_luadata/luadata.

blefloch commented on l3kernel/l3luatex.dtx:
```diff
@@ -370,9 +370,10 @@ local scan_int     = token.scan_int or token.scan_integer
-local false_tok    = token.create'prg_return_false:'
+local true_tok     = token_create'prg_return_true:'
+local false_tok    = token_create'prg_return_false:'
 %    \end{macrocode}
```
`true_tok` and `false_tok` seem unused.  Besides, don't they end up being undefined cs for the same reason as you explained for `use_none` and `use_i`?

zauguin replied:
They are used in `l3token` and defined here since they were considered to be generally useful. They would be undefined, but `l3bootstrap` explicitly makes sure that they appear to avoid that.

blefloch replied:
Should we have a wrapper like (untested code, and doesn't deal correctly with crazy category codes; it should `tex.print` some of it with the `-2` catcodetable)
```
token_create_safe = function(s)
  tex.runtoks(function()
    tex.sprint("\\edef\\__luatex_dummy:{\z
        \\expandafter\\noexpand\\csname" .. s .. "\\endcsname\z
    }\\endgroup")
  end)
  return token_create(s)
end
```

zauguin replied:
I like the idea, but would probably implement it based on `token.set_char` or so to avoid having to deal with catcode issues.

blefloch replied:
I don't understand how to avoid leaving the weird meaning in place on the TeX side when using `token.set_char`.  Is there a `token.set_undefined` or similar (I found nothing in the manual)?

zauguin replied:
There isn't. The idea would be to first run `\begingroup`, then `token.set_char` and finally `\endgroup`. That's a bit of a pain, but since other approaches probably require groups too it seems reasonable.

zauguin replied:
I added a `token_create_safe`  based on`set_char` and `let`.

blefloch replied:
Just looked at the code for `token_create_safe`.  It's really neat!

blefloch commented on l3kernel/l3luatex.dtx:
```diff
@@ -677,6 +677,85 @@ end
+% The actual work is done in \texttt{pre_dump}. The \texttt{luadata_order} is used
+% to ensure that the order is consistent over multiple runs.
+%    \begin{macrocode}
+luatexbase.add_to_callback("pre_dump", function()
```
Wrong indent.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,454 @@
+%    \end{macrocode}
+% \end{variable}
+%
+% \begin{macro}{\s_@@}
```
Given that `\s_@@` is not actually a scan mark (which are equal to `\relax`), it should probably be called something else like the `\@@_chk:w` used in `l3fp`.

zauguin replied:
Maybe just `\__intarray:w`?

blefloch replied:
Ok, if you prefer.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,454 @@
+
+local scan_token = token.scan_token
+local scan_toks = token.scan_toks
+local get_csname = token.get_csname
```
Seems `scan_toks` and `get_csname` are unused.  Same for `scan_argument` later on.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,454 @@
+  put_next(expand_after_scan_stop)
+  scan_token()
+  sprint(-2, item or 0)
+end, 'global')
```
This is not suitable yet in the false branch: because f-expansion gives `0\msg_expandable_error...` instead of producing the error first.  One idea is to avoid `sprint(0)` in the error case and simply put a `0` on the TeX side after the error.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+    scan_int()
+    put_next(use_i)
+  end
+end, 'protected', 'global')
```
In fact `@@_gset:w` should simply do no check at all I think, directly `local i=scan_int() @@_table()[i]=scan_int()`.  Then it can be used in some functions like setting from a clist, below.

zauguin replied:
Fixed

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,454 @@
+\cs_new_protected:Npn \intarray_const_from_clist:Nn #1#2
+  {
+    \@@_new:N #1
+    \@@_gset_count:Nw #1 \clist_count:n {#2} \scan_stop:
```
This step is wasteful and can be simply omitted I think provided `\__kernel_intarray_gset:Nnn` just does `t[i]=...` with no tests, extending the table as needed.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,454 @@
+    \int_value:w \int_eval:w #2 \exp_after:wN ;
+    \int_value:w \int_eval:w #3
+    #1
+  }
```
Perhaps just `\@@_range_to_clist:w #1 \int_eval:n {#2} ~ \int_eval:n {#3} ~` and then below `sprint(-2, concat(@@_table(), ',', scan_int(), scan_int()))`.

blefloch commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,454 @@
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\intarray_show:N, \intarray_show:c, \intarray_log:N, \intarray_log:c}
```
The definition of `\intarray_show:N` etc seems identical in all engines, should be shared.

zauguin replied:
The definition of `\@@_show:NN` will probably be slightly different once the second space at the beginning is fixed, therefore it has been kept separate. The `\intarray_show:N` etc. could still be merged but then `\@@_show:NN` would need separate documentation which would effectively just be a textual version of the definition of `\intarray_show:N` and `\intarray_log:N`, so it seemed useful to keep them here too.

blefloch replied:
I think the fix should be in `\@@_to_clist:Nn` for non-LuaTeX, such as adding an argument `#4` to `\@@_to_clist:w`, which is initially `\use_none:n` (when called from the main `_to_clist:Nn` and later is `\exp_not:n` (when called from `_to_clist:w`).  Many other ways obviously.

zauguin replied:
Sounds reasonable, so I moved the `show` functions to the shared part.

zauguin commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+%
+%    \begin{macrocode}
+%<*tex>
+\cs_generate_variant:Nn \intarray_count:N { c }
```
I tried moving it, but it feels rather artificial. I feel like the value of keeping the definition of `\intarray_count:N` and `\intarray_count:c` together seems higher than the value of avoiding duplication here. (Of course the same question appears in other cases of `\cs_generate_variant:Nn`.

blefloch replied:
You're right actually, sorry.

zauguin commented on l3kernel/l3intarray.dtx:
```diff
@@ -173,7 +173,529 @@
+%
+% Go to the false branch of the conditional above.
+%    \begin{macrocode}
+}{
```
Done

josephwright commented:
We good-to-go here now?

blefloch commented:
As far as I can tell, @zauguin has applies all suggested changes, and all conversations are resolved, so we should be good to merge if he is ok with it.
